pkgs/applications/misc/dmenu-rs/default.nix

```diff
@@ -0,0 +1,96 @@
+{ stdenv
+, rustPlatform
+, lib
+, fetchFromGitHub
+, cargo
+, clang
+, expat
+, fontconfig
+, libXft
+, libXinerama
+, m4
+, pkg-config
+, python3
+}:
+
+# The dmenu-rs package has extensive plugin support. However, this derivation
+# only builds the package with the default set of plugins. If you'd like to
+# further customize dmenu-rs you can build it from the source.
+# See: https://github.com/Shizcow/dmenu-rs#plugins
+stdenv.mkDerivation rec {
+  pname = "dmenu-rs";
+  version = "5.5.1";
+
+  src = fetchFromGitHub {
+    owner = "Shizcow";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-WpDqBjIZ5ESnoRtWZmvm+gNTLKqxL4IibRVCj0yRIFM=";
+  };
+
+  nativeBuildInputs = [
+    cargo
+    clang
+    m4
+    pkg-config
+    python3
+    rustPlatform.bindgenHook
+    rustPlatform.cargoSetupHook
+  ];
+
+  buildInputs = [
+    expat
+    fontconfig
+    libXft
+    libXinerama
+  ];
+
+  # The dmenu-rs repository does not include a Cargo.lock because of its
+  # dynamic build and plugin support. Generating it with make and checking it
+  # in to nixpkgs here was the easiest way to supply it to rustPlatform.
+  # See: https://github.com/Shizcow/dmenu-rs/issues/34#issuecomment-757415584
+  cargoDeps = rustPlatform.importCargoLock {
+    lockFile = ./Cargo.lock;
+  };
+
+  # The rust-xcb dependency dynamically generates rust code at build time.
+  # This derivation uses nixpkgs rust functions that vendor each cargo
+  # dependency's source code into the READ-ONLY nix store. To avoid the code
+  # generation step failing, we copy the rust-xcb source out of the nix store
+  # and make it writeable. Also, we remove the build's hardcoded c compiler.
+  # See: https://github.com/rust-x-bindings/rust-xcb/tree/v0.8.2
+  patchPhase = ''
+    runHook prePatch
+
+    substituteInPlace config.mk --replace "clang" ""
+
+    chmod +w "$NIX_BUILD_TOP/cargo-vendor-dir"
+    mkdir -p "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2-readwrite"
+    cp -r --no-preserve=mod "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2/." "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2-readwrite"
+    unlink "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2"
+    mv "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2-readwrite" "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2"
+
+    runHook postPatch
+  '';
+
+  postPatch = ''
```

Comment:
there is no reason to have both postPatch and a patchPhase. As the default patchPhase is still useful if someone downstream wants to add a patch with overrideAttrs, it is probably better to move the content of patchPhase (except runHook) to postPatch.

Reply:
OK. I've moved everything over into `postPatch`.

But, any guidance on what is a good definition of a patch is / what should generally go in `patchPhase`?

Should it only be applying `.patch` files? Or any change to the project's source code? Any change to the project's build? Any change to the project's dependencies' source code or build?

I guess what this `patchPhase` is currently doing is
  - changing the project's build code (`config.mk` and copying in `Cargo.lock`)
  - subtly change a vendored dependency so that it can be written to, without actually altering any of that dependency's source code or build code

So if drew the line at "patches are things that alter a project's source code (but not necessarily build code)" then it seems like it makes sense to keep it all out of `patchPhase`.

Probably there's no hard rule but just trying to get a feel for what is what in nixpkgs.

Reply:
> But, any guidance on what is a good definition of a patch is / what should generally go in `patchPhase`?
> 
> Should it only be applying `.patch` files? Or any change to the project's source code? Any change to the project's build? Any change to the project's dependencies' source code or build?

Typically, many distros (if not all!) provide `.patch` files in order to directly modify the source code of the apps they provide. 

A typical case is when a new compiler release does not compile some old code (indeed it happened to me when I tried to compile qemu in Slackware a long time ago...). So, usually a package manager provides patch files for cases like these.
Indeed, many times those patchsets are generated by merely diffing the `master` git branch with the current `stable` git branch, maybe because some people like ultra fresh, pre-alpha code...

Another thing that such package managers can do too is run sanitization scripts that clean up some quirks on the original source code; things like removing annoying lines from a Makefile or fixing hardcoded locations (e.g. FreeBSD ports system installs things on `/usr/local/` instead of `/usr/`; so maybe some tool that expects to find things on /usr will not get it without help).

It gets a bit complicated when both things need to be done: `.patch` files to modify C source code, plus sanitization scripts that cleanup Makefiles.

Now, switch to Nix(pkgs).

In Nixpkgs, we provide both use cases via the same `patchPhase`. However, as I said above, it is a bit complicated...

The typical, unedited patchPhase usually works like this:

```nix
patchPhase = ''
  runHook prePatch

  patch -p1 < <the .patch files provided in patches =[ ... ];>
 
  runHook postPatch
''
```

The `runHook`s are there because of composability:

Nix has the ability of compose overlays of package trees, in a similar manner when you add third party repositories in other package managers (`/etc/apt/sources.lst` anyone?).
Such pre and post hooks allow the creator of an overlay to easily modify the phases without having to rewrite them completely.

Well, when you customize this phase as above, it would look like

```nix
patchPhase = ''
  runHook prePatch

  <a giant script modifying the source code>
 
  runHook postPatch
''
```

However, by doing this, you are basically ignoring any `.patch` files eventually provided in `patches` list.

So, when you need to use such "sanitization scripts", it is better to use `postPatch` instead of rewriting the whole `patchPhase`.

> Probably there's no hard rule but just trying to get a feel for what is what in nixpkgs.

Sometimes many things are developed by trial and error.

Reply:
Ah thank you for all that good background knowledge!